crates/sc-subspace-block-relay/src/lib.rs

```diff
@@ -153,3 +159,32 @@ struct ProtocolUnitInfo<ProtocolUnitId, ProtocolUnit> {
     /// action is needed on client side to resolve it
     unit: Option<ProtocolUnit>,
 }
+
+/// The relay protocol identifier.
+#[derive(Copy, Clone, Eq, PartialEq, Encode, Decode)]
+pub(crate) enum RelayProtocol {
+    #[codec(index = 0)]
+    CompactBlock,
+    // Next protocol goes here.
+    // #[codec(index = 1)]
+}
+
+/// The relay protocol version.
+#[derive(Copy, Clone, Eq, PartialEq, Encode, Decode)]
+pub(crate) struct RelayVersion {
```

Comment:
`RelayVersion` and `RelayProtocol` are more confusing to me TBH. Looking at the current implementation of `sc-block-relay`, why not define a new type that encapsulates different protocols? Something like this maybe
```
#[derive(Encode, Decode)]
enum ProtocolRequest<CompactBlockRequest, GrapheneBlockRequest> {
	CompactBlockReq(CompactBlockRequest)
	GrapheneBlockReq(GrapheneBlockRequest)
}
```

As for the version of the the protocol, i dont think protocol needs to leak that information. I would personally version the specific protocol request from the get go if there will be new versions coming up. Something like this
```
#[derive(Encode, Decode)]
enum CompactBlockRequest {
	V0 (CompactBlockRequestV0)
	V1 (CompactBlockRequestV1)
}
```

This would eliminate bunch of the protocol version checks and protocol either reponds to the request it supports or dont.

Reply:
I was thinking about maybe not even abstracting over requests and responses, have them be just "messages" and let the protocol implementation decide what it is internally.

Reply:
> I was thinking about maybe not even abstracting over requests and responses, have them be just "messages" and let the protocol implementation decide what it is internally.

Ahh since there will be only one protocol at a time ? I was under the assumption that we will have multiple protocols a server supports at a time.

Reply:
For multiple protocols:
```rust
#[derive(Encode, Decode)]
enum ProtocolMessage {
    CompactBlock(CompactBlockProtocolMessage)
    Graphene(GrapheneProtocolMessage)
}

#[derive(Encode, Decode)]
enum CompactBlockProtocolMessage {
    Request(CompactBlockRequest)
    Response(CompactBlockResponse)
}
```

Reply:
No, this is exactly the I want to avoid (also mentioned in https://github.com/subspace/subspace/pull/1937#discussion_r1326550384), about creating global enum and breaking the protocol abstraction